[AArch64][SelectionDAG] Vector splitting and promotion for histogram intrinsic

DevM-uk opened the pull request:
Adds support for wider-than-legal vector types for the histogram intrinsic (llvm.experimental.vector.histogram.add) by splitting the vector.

DevM-uk committed:
[AArch64] Implement vector splitting for histogram intrinsic

Adds support for wider-than-legal vector types for the histogram intrinsic (llvm.experimental.vector.histogram.add) by splitting the vector.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -25434,6 +25436,41 @@ performScalarToVectorCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
   return NVCAST;
 }
 
+static SDValue performHistogramCombine(SDNode *N,
```
Rather than doing this as a target-specific DAG combine, this should be done as generic type legalisation in LegalizeVectorTypes.cpp. See `DAGTypeLegalizer::SplitVectorOperand` for how this is done for other operations.

DevM-uk replied:
You're right. I've moved the splitting to `DAGTypeLegalizer::SplitVecOp_VECTOR_HISTOGRAM` and added/moved the `INC` operand promotion to `DAGTypeLegalizer::PromoteIntOp_VECTOR_HISTOGRAM`.

DevM-uk committed:
Move legalization to legalizer

Moves the splitting from a target specific DAG combine to LegalizeVectorTypes and adds promotion to LegalizeIntegerTypes. Also updates how the histogram intrinsic is custom lowered during legalize ops.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
+SDValue DAGTypeLegalizer::SplitVecOp_VECTOR_HISTOGRAM(SDNode *N) {
+  MaskedHistogramSDNode *HG = cast<MaskedHistogramSDNode>(N);
+  SDLoc DL(HG);
+  SDValue Chain = HG->getChain();
```
nit: `Chain` has one use, just inline.

DevM-uk replied:
Done.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
+  SDValue Inc = HG->getInc();
+  SDValue Mask = HG->getMask();
+  SDValue Ptr = HG->getBasePtr();
+  SDValue Index = HG->getIndex();
```
nit: `Index` has one use, just inline.

DevM-uk replied:
Done.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
+  SDLoc DL(HG);
+  SDValue Chain = HG->getChain();
+  SDValue Inc = HG->getInc();
+  SDValue Mask = HG->getMask();
```
nit: `Mask` has one use, just inline.

DevM-uk replied:
Done.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
+  SDValue IndexLo, IndexHi, MaskLo, MaskHi;
+  std::tie(IndexLo, IndexHi) = DAG.SplitVector(Index, DL);
+  std::tie(MaskLo, MaskHi) = DAG.SplitVector(Mask, DL);
+  SDValue HistogramOpsLo[] = {Chain, Inc, MaskLo, Ptr, IndexLo, Scale, IntID};
```
nit:
```suggestion
  SDValue OpsLo[] = {Chain, Inc, MaskLo, Ptr, IndexLo, Scale, IntID};
```
or just inline the initializer-sequence into the `getMaskedHistogram` directly.

DevM-uk replied:
I've made the suggested change. Operand lists are passed in using arrays in other places (such as with `getMaskedScatter`s) rather than passing the initializer-sequence inline.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4274,6 +4277,31 @@ SDValue DAGTypeLegalizer::SplitVecOp_VP_CttzElements(SDNode *N) {
+  std::tie(IndexLo, IndexHi) = DAG.SplitVector(Index, DL);
+  std::tie(MaskLo, MaskHi) = DAG.SplitVector(Mask, DL);
+  SDValue HistogramOpsLo[] = {Chain, Inc, MaskLo, Ptr, IndexLo, Scale, IntID};
+  SDValue HChain = DAG.getMaskedHistogram(DAG.getVTList(MVT::Other), MemVT, DL,
```
nit:
```suggestion
  SDValue Lo = DAG.getMaskedHistogram(DAG.getVTList(MVT::Other), MemVT, DL,
```

DevM-uk replied:
Done.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1777,10 +1777,12 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
     if (Subtarget->hasSVE2()) {
-      setOperationAction(ISD::EXPERIMENTAL_VECTOR_HISTOGRAM, MVT::Other,
+      setOperationAction(ISD::EXPERIMENTAL_VECTOR_HISTOGRAM, MVT::i8, Promote);
+      setOperationAction(ISD::EXPERIMENTAL_VECTOR_HISTOGRAM, MVT::i16, Promote);
```
These two `setOperationActions`s are unnecessary, because i8 and i16 are illegal types, so the legalizer will already try to legalize the operation, regardless of this `setOperationAction`.

DevM-uk committed:
fixup: Remove unnecessary setOperationActions

sdesmalen-arm reviewed (approved):
Left a few nits, but otherwise looks good to me!

sdesmalen-arm commented:
On the commit message, the patch is also doing type promotion, so maybe you want to rephrase it to something slightly more generic, e.g. `Vector splitting and promotion for histogram` ?

DevM-uk committed:
fixup: Review nits

DevM-uk committed:
fixup: Remove unused variable
